File: NaoLinear/extended01.m
%Exemplo de sistema dinamico linear, mas com medicao nao-linear
clear all
close all

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% DADOS INICIAIS %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

sigma_acel = 0.2;                                                          %Constante de desvio da aceleracao 
sigma_medicao_posicao = 5;                                                 %Desvio da medicao de posicao
sigma_medicao_angulo = 0.0087;                                             %Desvio da medicao do angulo
delta_tempo = 1;                                                           %Variacao do tempo

%Matriz x 6x1 de estimativas
x = zeros(6, 1); 
x(1) = 400;                                                                %Posicao x inicial 
x(4) = -300;                                                               %Posicao y inicial

%Matriz F e transposta Ft de transicao de estados. Linear neste exemplo
F = [1, delta_tempo, (delta_tempo^2)/2, 0, 0, 0; 
     0, 1, delta_tempo, 0, 0, 0;
     0, 0, 1, 0, 0, 0;
     0, 0, 0, 1, delta_tempo, (delta_tempo^2)/2;
     0, 0, 0, 0, 1, delta_tempo;
     0, 0, 0, 0, 0, 1];
Ft = transpose(F);

%Matriz P 6x6 de covariancia de posicao
P = zeros(6,6);          
for i = 1:1:6           
    P(i,i) = 500;                                                          %Preenche a diagonal principal da matriz P
end

%Matriz Q de ruido da covariancia
Q = [(delta_tempo^4)/4, (delta_tempo^3)/2, (delta_tempo^2)/2, 0, 0, 0; 
     (delta_tempo^3)/2, delta_tempo^2, delta_tempo, 0, 0, 0;
     (delta_tempo^2)/2, delta_tempo, 1, 0, 0, 0;
     0, 0, 0, (delta_tempo^4)/4, (delta_tempo^3)/2, (delta_tempo^2)/2;
     0, 0, 0, (delta_tempo^3)/2, delta_tempo^2, delta_tempo;
     0, 0, 0, (delta_tempo^2)/2, delta_tempo, 1];
Q = Q*(sigma_acel^2);

%Matriz R de covariancia de ruido da medicao 
R = [sigma_medicao_posicao^2 0 ; 0 sigma_medicao_angulo^2]; 

I = eye(6);                                                                %Matriz identidade

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% ALGORITMO DE KALMAN %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

for i = 1:1:2
    x = F*x;                                                               %Predicao da estimativa
    
    P = F*P*Ft + Q;                                                        %Predicao da covariancia
    
    %Entrada da medicao do usuario
    z(1,1) = input(sprintf('Digite o valor da medicao z1: '));
    z(2,1) = input(sprintf('Digite o valor da medicao z2: '));

    %Funcao para calcular jacobiano de H
    [r, phi, H, dH, dHt] = jacobiano(x);
    
    K = P*dHt*(dH*P*dHt + R)^(-1);                                         %Atualizacao do Ganho de Kalman
    Kt = transpose(K);
    
    x = x + K*(z - H);                                                     %Atualizacao da estimativa
    
    P = (I - K*dH)*P*(transpose(I - K*dH)) + K*R*Kt;                       %Atualizacao da predicao
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% CALCULO DO JACOBIANO %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function [r, phi, H, dH, dHt] = jacobiano(x)
    %Calcula r e phi
    r = x(1)^2 + x(4)^2;
    phi = atan(x(4)/x(1));

    %Calcula H
    H = [sqrt(r);
         phi];

    %Calcula dH: jacobiano de H
    dH = [x(1)/sqrt(r) 0 0 x(4)/sqrt(r) 0 0 ;
          -x(4)/r 0 0 x(1)/r 0 0];

    %Matriz transposta do jacobiano de H
    dHt = transpose(dH);
end    

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

File: NaoLinear/C/extended01.c
#include <stdio.h>

#define ROWS 6                                                                                                          //Linhas das matrizes
#define COLUMNS 6                                                                                                       //Colunas das matriZES
#define X_COLUMNS 1                                                                                                     //Colunas da matriz x


void initialize_x(float x[ROWS][X_COLUMNS]){
    int i, j;

    for(i = 0; i < ROWS; i++){                                                                                          //Percorre linhas da matriz
        for(j = 0; j < X_COLUMNS; j++){                                                                                 //Percorre colunas da matriz
            x[i][j] = 0;                                                                                                //Preenche matriz com zeros
        }
    }

    x[0][0] = 400;                                                                                                      //Posicao x inicial
    x[0][3] = -300;                                                                                                     //Posicao y inicial

}

void initialize_F(float F[ROWS][COLUMNS], float delta_tempo){
    int i, j;

    for(i = 0; i < ROWS; i++){                                                                                          //Percorre linhas da matriz
        for(j = 0; j < COLUMNS; j++){                                                                                   //Percorre colunas da matriz
            F[i][j] = 0;                                                                                                //Preenche matriz com zeros
            if(i == j){
                F[i][j] = 1;                                                                                            //Preenche diagonal principal com 1
            }
        }
    }

    F[0][1] = delta_tempo;
    F[1][2] = delta_tempo;
    F[3][4] = delta_tempo;
    F[4][5] = delta_tempo;
    F[0][2] = (delta_tempo*delta_tempo)/2;
    F[3][5] = (delta_tempo*delta_tempo)/2;

}

void initialize_P(float P[ROWS][COLUMNS]){
    int i, j;

    for(i = 0; i < ROWS; i++){                                                                                          //Percorre linhas da matriz
        for(j = 0; j < COLUMNS; j++){                                                                                   //Percorre colunas da matriz
            P[i][j] = 0;                                                                                                //Preenche matriz com zeros
            if(i == j){
                P[i][j] = 500;                                                                                          //Preenche diagonal principal com 500
            }
        }
    }

}

initialize_Q(float Q[ROWS][COLUMNS], float delta_tempo, float sigma_acel){
    int i, j;

    for(i = 0; i < ROWS; i++){                                                                                          //Percorre linhas da matriz
        for(j = 0; j < COLUMNS; j++){                                                                                   //Percorre colunas da matriz
            Q[i][j] = 0;                                                                                                //Preenche matriz com zeros
        }
    }

    Q[0][0] = (delta_tempo * delta_tempo * delta_tempo * delta_tempo) / 4;
    Q[0][1] = (delta_tempo * delta_tempo * delta_tempo) / 2;
    Q[0][2] = (delta_tempo * delta_tempo) / 2;
    Q[1][0] = (delta_tempo * delta_tempo * delta_tempo) / 2;
    Q[1][1] = (delta_tempo * delta_tempo);
    Q[1][2] = delta_tempo;
    Q[2][0] = (delta_tempo * delta_tempo) / 2;
    Q[2][1] = delta_tempo;
    Q[2][2] = 1;
    Q[3][3] = (delta_tempo * delta_tempo * delta_tempo * delta_tempo) / 4;
    Q[3][4] = (delta_tempo * delta_tempo * delta_tempo) / 2;
    Q[3][5] = (delta_tempo * delta_tempo) / 2;
    Q[4][3] = (delta_tempo * delta_tempo * delta_tempo) / 2;
    Q[4][4] = (delta_tempo * delta_tempo);
    Q[4][5] = delta_tempo;
    Q[5][3] = (delta_tempo * delta_tempo) / 2;
    Q[5][4] = delta_tempo;
    Q[5][5] = 1;

    for (i = 0; i < ROWS; i++) {
        for (j = 0; j < COLUMNS; j++) {
            Q[i][j] *= (sigma_acel * sigma_acel);                                                                       //Multiplica Q por (sigma_acel)^2
        }
    }

}

int main() {
    float sigma_acel = 0.2;                                                                                             //Constante de desvio da aceleracao
    float sigma_medicao_posicao = 5;                                                                                    //Desvio da medicao de posicao
    float sigma_medicao_angulo = 0.0087;                                                                                //Desvio da medicao do angulo
    float delta_tempo = 1;                                                                                              //Variacao do tempo

    //Matriz x de estimativas
    float x[ROWS][X_COLUMNS];                                                                                           //Matriz x de estimativas
    initialize_x(x);                                                                                                    //Inicializa a matriz x com os valores

    float F[ROWS][COLUMNS];                                                                                             //Matriz F de transicao de estados
    initialize_F(F, delta_tempo);                                                                                       //Inicializa a matriz F com os valores
    //FAZER transpose(F)

    float P[ROWS][COLUMNS];                                                                                             //Matriz P de covariancia
    initialize_P(P);                                                                                                    //Inicializa a matriz P com os valores

    float Q[ROWS][COLUMNS];                                                                                             //Matriz Q de ruido da covariancia
    initialize_Q(Q, delta_tempo, sigma_acel);                                                                           //Inicializa a matriz Q com os valores

    return 0;
}
